Write digits in groups of eight

TobiSchluter opened the pull request:
Writes the digits in blocks of eight instead of two and then adjust for the triailing zeros only once. They are identified by bit-counting of the eight-digit block.

This is an extension of my previous pull request. Unlike the previous, this benchmarks on par on linux and avoids all issues due to bit counting builtins not being clearly specified. This is achieved by relying on the compiler to find the right instruction. I explored `std::countl_zero` but it seemed to pessimize performance, at least on Windows.

On Windows it generally benchmarks minimally faster, on Linux it benchmarks on par, and maybe faster when `-mlzcnt` is given which allows the more efficient LZCNT instruction to be used (unlike BSR it needs no special handling of the zero case).

It removes two branches compared to the main branch, but adds two unconditional multiplications in the case of short decimal representations (which are unlikely, because a zero digit is less probable than a non-zero digit).

Here is a "lucky" benchmark run with `kTrial == 20` on Linux, generally the relative performance fluctuates a little. `zmjj_new` is this version
<img width="1615" height="1202" alt="image" src="https://github.com/user-attachments/assets/0c0b9f18-4cda-4b61-a6b3-d0e9ffd60303" />

This version clocks in at <s>322</s>320 lines of assembly <s>https://godbolt.org/z/5YTzY3zEb</s>https://godbolt.org/z/f3vv1Y3f7 (updated for second patch).
The current main branch compiles to 339 lines of assembly, including an additional 100byte buffer https://godbolt.org/z/exr3E581E

vitaut commented:
Thanks for the PR! I ran a quick test and the performance looks nearly identical to the current version with the lookup table on macOS with Apple M1.

vitaut commented on zmij.cc:
```diff
@@ -736,13 +782,13 @@ inline void write2digits(char* buffer, uint32_t value) noexcept {
+  write2digits(&bits.b[2], bb);
+  write2digits(&bits.b[4], cc);
+  write2digits(&bits.b[6], dd);
+  return bits.u64;
```
I'm pretty sure this is a UB in C++.

TobiSchluter replied:
Me too! I was looking for a reasonably clean form that guarantees that the value is materialized in a register, and couldn't find an option that doesn't look like
   memset((char *)some_uint64 + 2, digits(bb), 2)
which I'm fairly sure is undefined as well, no matter how I precisely address the bit range in the uint64_t. Anyway, I think I came up with something that works and is not actually undefined, I just need to confirm that the assembly looks reasonable.

TobiSchluter replied:
I changed this to use memcpy to uint16_t variables which are then shifted and or'ed to yield the final uint64_t. The assembly ended up becoming a bit shorter as well.

vitaut commented on CMakeLists.txt:
```diff
@@ -12,5 +12,6 @@ enable_testing()
 add_subdirectory(gtest)
 
 add_executable(zmij-test zmij-test.cc)
+target_compile_features(zmij-test PRIVATE cxx_std_20)
```
This looks unused.

vitaut replied:
We only need C++17. I removed the only C++20 feature (bit_cast) earlier.

TobiSchluter replied:
well sure, then it's C++17, I missed that change.

vitaut commented on zmij.cc:
```diff
@@ -757,28 +803,29 @@ auto write_significand(char* buffer, uint64_t value) noexcept -> char* {
+  uint64_t buf = write8digits(bb, cc, dd, ee);
+  memcpy(buffer, &buf, 8);
+
+  if (ffgghhii == 0) [[unlikely]] {
```
Why unlikely?

TobiSchluter replied:
There are 10^8 possible combinations for the trailing 8 digits but only one of those combinations takes that path.

That is of course an approximate statement given that not every decimal is representable as binary, but I assume it's close enough.

The actual distribution is probably also skewed by user input which will tend shorter. But the case where the user prints their manual input is probably not the most performance-critical?

vitaut replied:
I'm not sure we can make such assumptions about the distribution of user data. It is very much possible that all of it is rounded or something like that. In any case I don't think it has significant effect so I just removed the attribute for now.

TobiSchluter commented:
I force-pushed a second time, because I think it's easier to review as a single commit. Sorry about the noise.

Since I didn't keep the commit messages, let me just say that there are two changes since the review:
1. I changed the string length evaluation to avoid the zero length check even if the compiler emits BSR (which gcc does by default on x64)
2. I renamed the variables that store the 8byteblocks from buf, buf2 to bits, bits2. Still not happy with the name, but there is no problem harder in programming than naming.
Edit:
3. I also accidentally removed the change which added C++20 mode for zmij-test. I still need it on MSVC.

vitaut commented:
BTW Paul Khuong shared his post https://pvk.ca/Blog/2017/12/22/appnexus-common-framework-its-out-also-how-to-print-integers-faster/. Looks like he used a similar technique for itoa and came to conclusion that "using registers as buffers while generating digits is a win".

TobiSchluter commented:
ps reading your blog post made me think how a more compact table-based approach would work. Basically, there are many unused bits in the digits2::data array. So I tried merging the length information into the digits array, doing something like this:
```
struct bits_and_len {
    uint16_t bits;
    size_t len;
};

inline auto digits2_len(size_t value) noexcept -> bits_and_len {
  constexpr uint16_t one = 0x4000;
  constexpr uint16_t two = 0x8000;
  constexpr uint16_t mask = 0xC000;
  constexpr size_t shift = 15;
  static const uint16_t data[] = {
    0x3030 | two, 0x3130, 0x3230, 0x3330, 0x3430, 0x3530, 0x3630, 0x3730, 0x3830, 0x3930,
    0x3031 | one, 0x3131, 0x3231, 0x3331, 0x3431, 0x3531, 0x3631, 0x3731, 0x3831, 0x3931,
    0x3032 | one, 0x3132, 0x3232, 0x3332, 0x3432, 0x3532, 0x3632, 0x3732, 0x3832, 0x3932,
    0x3033 | one, 0x3133, 0x3233, 0x3333, 0x3433, 0x3533, 0x3633, 0x3733, 0x3833, 0x3933,
    0x3034 | one, 0x3134, 0x3234, 0x3334, 0x3434, 0x3534, 0x3634, 0x3734, 0x3834, 0x3934,
    0x3035 | one, 0x3135, 0x3235, 0x3335, 0x3435, 0x3535, 0x3635, 0x3735, 0x3835, 0x3935,
    0x3036 | one, 0x3136, 0x3236, 0x3336, 0x3436, 0x3536, 0x3636, 0x3736, 0x3836, 0x3936,
    0x3037 | one, 0x3137, 0x3237, 0x3337, 0x3437, 0x3537, 0x3637, 0x3737, 0x3837, 0x3937,
    0x3038 | one, 0x3138, 0x3238, 0x3338, 0x3438, 0x3538, 0x3638, 0x3738, 0x3838, 0x3938,
    0x3039 | one, 0x3139, 0x3239, 0x3339, 0x3439, 0x3539, 0x3639, 0x3739, 0x3839, 0x3939,
  };

  return { .bits = uint16_t(data[value] & ~mask), .len = unsigned(data[value]) >> shift};
}
```

No impact on performance, but somehow it generates significantly more assembly. At a quick perusal I can't tell why, it doesn't seem to be the obvious suspect: no excessive bit operations in sight.

TobiSchluter commented on zmij.cc:
```diff
@@ -710,13 +710,59 @@ inline auto digits2(size_t value) noexcept -> const char* {
+  return _lzcnt_u64(x);
+}
+#else
+// Both clang and gcc recognize this implementation and replace it with the
```
My bad, I had updated the comment to say
```
// Unlike MSVC, clang and gcc recognize this implementation
```
but I should have said so explicitly over here.

Here's a godbolt to that effect
https://godbolt.org/z/Waxj54v7W

It's also of note that as of now gcc generates the zero-check even though it uses the lzcnt instruction. I'm not sure what the circumstances are that make it do that, maybe it's dependent on -march settings.

TobiSchluter commented:
> BTW Paul Khuong shared his post https://pvk.ca/Blog/2017/12/22/appnexus-common-framework-its-out-also-how-to-print-integers-faster/. Looks like he used a similar technique for itoa and came to conclusion that "using registers as buffers while generating digits is a win".

Interesting! I was thinking that this could be done, but didn't think it through myself. I expanded this approach to eight simultaneous digits, and it passes the godbolt test. Still need to benchmark it and get rid of some repeated masking and unmasking of 0x30,

xjb714 commented:
@TobiSchluter In fact, on a little-endian machine, we can use the following method to calculate the BCD code of integers within the range [0, 1e8). x is the input.
```
        u64 aabb_ccdd_merge = (x << 32) - ((10000ull<<32) - 1) * ((x * 109951163) >> 40);
        u64 aa_bb_cc_dd_merge = (aabb_ccdd_merge << 16) - ((100ull<<16) - 1) * (((aabb_ccdd_merge * 10486) >> 20) & ((0x7FULL << 32) | 0x7FULL));
        u64 aabbccdd_BCD = (aa_bb_cc_dd_merge << 8) - ((10ull<<8) - 1) * (((aa_bb_cc_dd_merge * 103) >> 10) & ((0xFULL << 48) | (0xFULL << 32) | (0xFULL << 16) | 0xFULL));
```

xjb714 commented:
This code is used on a big-endian CPU.

```cpp
    u64 aabb_ccdd_merge = aabbccdd + ((1ull<<32) - 10000) * ((aabbccdd * 109951163) >> 40);
    u64 aa_bb_cc_dd_merge = aabb_ccdd_merge + ((1ull<<16) - 100) * ((aabb_ccdd_merge * 10486) >> 20) & ((0x7FULL << 32) | 0x7FULL);
    u64 aabbccdd_BCD = aa_bb_cc_dd_merge + ((1ull<<8) - 10) * ((aa_bb_cc_dd_merge * 103) >> 10) & ((0xFULL << 48) | (0xFULL << 32) | (0xFULL << 16) | 0xFULL);
```

TobiSchluter commented:
> @TobiSchluter In fact, on a little-endian machine, we can use the following method to calculate the BCD code of integers within the range [0, 1e8). x is the input.
> 
> ```
>         u64 aabb_ccdd_merge = (x << 32) - ((10000ull<<32) - 1) * ((x * 109951163) >> 40);
>         u64 aa_bb_cc_dd_merge = (aabb_ccdd_merge << 16) - ((100ull<<16) - 1) * (((aabb_ccdd_merge * 10486) >> 20) & ((0x7FULL << 32) | 0x7FULL));
>         u64 aabbccdd_BCD = (aa_bb_cc_dd_merge << 8) - ((10ull<<8) - 1) * (((aa_bb_cc_dd_merge * 103) >> 10) & ((0xFULL << 48) | (0xFULL << 32) | (0xFULL << 16) | 0xFULL));
> ```

That's a fun little trick to combine div and mod. I'll see if the compiler can figure it out by itself though, because that's definitely not easy to read 😆
